In [1]:
import pandas as pd

In [2]:
real_estate = pd.read_csv('../00-data/Real estate.csv')
real_estate.columns=['ID', 'transaction_date', 'house_age', 'nearest_MRT_station_distance', 'minimarket_nr', 'lat', 'long', 'price_m2']

real_estate

ID  transaction_date  house_age  nearest_MRT_station_distance  \
0      1          2012.917       32.0                      84.87882   
1      2          2012.917       19.5                     306.59470   
2      3          2013.583       13.3                     561.98450   
3      4          2013.500       13.3                     561.98450   
4      5          2012.833        5.0                     390.56840   
..   ...               ...        ...                           ...   
409  410          2013.000       13.7                    4082.01500   
410  411          2012.667        5.6                      90.45606   
411  412          2013.250       18.8                     390.96960   
412  413          2013.000        8.1                     104.81010   
413  414          2013.500        6.5                      90.45606   

     minimarket_nr       lat       long  price_m2  
0               10  24.98298  121.54024      37.9  
1                9  24.98034  121.53951      42.2  
2                5  24.98746  121.54391      47.3  
3                5  24.98746  121.54391      54.8  
4                5  24.97937  121.54245      43.1  
..             ...       ...        ...       ...  
409              0  24.94155  121.50381      15.4  
410              9  24.97433  121.54310      50.0  
411              7  24.97923  121.53986      40.6  
412              5  24.96674  121.54067      52.5  
413              9  24.97433  121.54310      63.9  

[414 rows x 8 columns]

In [3]:
north_east_point = (real_estate['lat'].max(), real_estate['long'].max())
north_east_point

(np.float64(25.01459), np.float64(121.56627))

In [4]:
import numpy as np

real_estate_ne = real_estate.copy()
real_estate_ne['distance_from_NE'] = np.sqrt(
    (north_east_point[0] - real_estate_ne['lat'])**2 +
    (north_east_point[1] - real_estate_ne['long'])**2
)

real_estate_ne

ID  transaction_date  house_age  nearest_MRT_station_distance  \
0      1          2012.917       32.0                      84.87882   
1      2          2012.917       19.5                     306.59470   
2      3          2013.583       13.3                     561.98450   
3      4          2013.500       13.3                     561.98450   
4      5          2012.833        5.0                     390.56840   
..   ...               ...        ...                           ...   
409  410          2013.000       13.7                    4082.01500   
410  411          2012.667        5.6                      90.45606   
411  412          2013.250       18.8                     390.96960   
412  413          2013.000        8.1                     104.81010   
413  414          2013.500        6.5                      90.45606   

     minimarket_nr       lat       long  price_m2  distance_from_NE  
0               10  24.98298  121.54024      37.9          0.040948  
1                9  24.98034  121.53951      42.2          0.043464  
2                5  24.98746  121.54391      47.3          0.035157  
3                5  24.98746  121.54391      54.8          0.035157  
4                5  24.97937  121.54245      43.1          0.042519  
..             ...       ...        ...       ...               ...  
409              0  24.94155  121.50381      15.4          0.096105  
410              9  24.97433  121.54310      50.0          0.046451  
411              7  24.97923  121.53986      40.6          0.044134  
412              5  24.96674  121.54067      52.5          0.054268  
413              9  24.97433  121.54310      63.9          0.046451  

[414 rows x 9 columns]

In [5]:
output = ['price_m2']
X = real_estate.drop(output, axis=1).values
y = real_estate[output].values

In [6]:
print(X)
print(y)

[[1.0000000e+00 2.0129170e+03 3.2000000e+01 ... 1.0000000e+01
  2.4982980e+01 1.2154024e+02]
 [2.0000000e+00 2.0129170e+03 1.9500000e+01 ... 9.0000000e+00
  2.4980340e+01 1.2153951e+02]
 [3.0000000e+00 2.0135830e+03 1.3300000e+01 ... 5.0000000e+00
  2.4987460e+01 1.2154391e+02]
 ...
 [4.1200000e+02 2.0132500e+03 1.8800000e+01 ... 7.0000000e+00
  2.4979230e+01 1.2153986e+02]
 [4.1300000e+02 2.0130000e+03 8.1000000e+00 ... 5.0000000e+00
  2.4966740e+01 1.2154067e+02]
 [4.1400000e+02 2.0135000e+03 6.5000000e+00 ... 9.0000000e+00
  2.4974330e+01 1.2154310e+02]]
[[ 37.9]
 [ 42.2]
 [ 47.3]
 [ 54.8]
 [ 43.1]
 [ 32.1]
 [ 40.3]
 [ 46.7]
 [ 18.8]
 [ 22.1]
 [ 41.4]
 [ 58.1]
 [ 39.3]
 [ 23.8]
 [ 34.3]
 [ 50.5]
 [ 70.1]
 [ 37.4]
 [ 42.3]
 [ 47.7]
 [ 29.3]
 [ 51.6]
 [ 24.6]
 [ 47.9]
 [ 38.8]
 [ 27. ]
 [ 56.2]
 [ 33.6]
 [ 47. ]
 [ 57.1]
 [ 22.1]
 [ 25. ]
 [ 34.2]
 [ 49.3]
 [ 55.1]
 [ 27.3]
 [ 22.9]
 [ 25.3]
 [ 47.7]
 [ 46.2]
 [ 15.9]
 [ 18.2]
 [ 34.7]
 [ 34.1]
 [ 53.9]
 [ 38.3]
 [ 42. ]
 [ 61.5]
 [ 1

In [7]:
from sklearn.model_selection import train_test_split

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.3)

print(f"{X_train.shape}, {y_train.shape}, {X_test.shape}, {y_test.shape}")

(289, 7), (289, 1), (125, 7), (125, 1)


In [9]:
from sklearn.preprocessing import StandardScaler, PolynomialFeatures


In [10]:
ss = StandardScaler()
X_train_std = ss.fit_transform(X_train)
X_test_std = ss.transform(X_test)

In [11]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [12]:
scores = []

for i in range(1,11):
    pf = PolynomialFeatures(degree=i)
    X_train_poly = pf.fit_transform(X_train_std)
    X_test_poly = pf.fit_transform(X_test_std)
    ll = LinearRegression()
    ll.fit(X_train_poly, y_train)
    y_pred_train = ll.predict(X_train_poly)
    y_pred_test = ll.predict(X_test_poly)
    print(f'poly: {i}')
    print(f'TRAIN -> mse: {mean_squared_error(y_train, y_pred_train)}; r2: {r2_score(y_train, y_pred_train)}')
    print(f'TEST -> mse: {mean_squared_error(y_test, y_pred_test)}; r2: {r2_score(y_test, y_pred_test)}')
    print('-----------------------------------------------------')

poly: 1
TRAIN -> mse: 78.74157321400966; r2: 0.5888965057262947
TEST -> mse: 74.70865668115285; r2: 0.5532450005316663
-----------------------------------------------------
poly: 2
TRAIN -> mse: 53.994270128325006; r2: 0.7181002078764228
TEST -> mse: 57.87816036064387; r2: 0.6538907450644771
-----------------------------------------------------
poly: 3
TRAIN -> mse: 39.26387868597756; r2: 0.7950064106203006
TEST -> mse: 81.26852218854351; r2: 0.5140172477991507
-----------------------------------------------------
poly: 4
TRAIN -> mse: 8.59349711928002e-23; r2: 1.0
TEST -> mse: 684888.0277886384; r2: -4094.6050353903865
-----------------------------------------------------
poly: 5
TRAIN -> mse: 1.6381681653900607e-24; r2: 1.0
TEST -> mse: 9790.570278531895; r2: -57.5472475866863
-----------------------------------------------------
poly: 6
TRAIN -> mse: 8.401522525802366e-24; r2: 1.0
TEST -> mse: 444520.29915323056; r2: -2657.214922260339
-----------------------------------------------

ho un overfitting molto marcato dal quarto in poi, inizio rimuovendo lat e long

versione con tutti i valori
poly 1 => scarso
poly 2 => accettabile
poly 3 => inizia overfitting

In [13]:
to_remove = ['price_m2', 'lat', 'long']
X_removed = real_estate.drop(to_remove, axis=1).values


In [14]:
X_train_removed, X_test_removed, y_train_removed, y_test_removed = train_test_split(X_removed, y, test_size=0.3, random_state=42)

print(f"{X_train_removed.shape}, {X_test_removed.shape}, {y_train_removed.shape}, {y_test_removed.shape}")

(289, 5), (125, 5), (289, 1), (125, 1)


In [15]:
ss_removed = StandardScaler()
X_train_removed_std = ss_removed.fit_transform(X_train_removed)
X_test_removed_std = ss_removed.transform(X_test_removed)

In [16]:
scores_removed = []

for i in range(1,11):
    pf_removed = PolynomialFeatures(degree=i)
    X_train_removed_poly = pf_removed.fit_transform(X_train_removed_std)
    X_test_removed_poly = pf_removed.fit_transform(X_test_removed_std)
    ll_removed = LinearRegression()
    ll_removed.fit(X_train_removed_poly, y_train_removed)
    y_pred_removed_train = ll_removed.predict(X_train_removed_poly)
    y_pred_removed_test = ll_removed.predict(X_test_removed_poly)
    print(f'poly: {i}')
    print(f'TRAIN -> mse: {mean_squared_error(y_train_removed, y_pred_removed_train)}; r2: {r2_score(y_train_removed, y_pred_removed_train)}')
    print(f'TEST -> mse: {mean_squared_error(y_test_removed, y_pred_removed_test)}; r2: {r2_score(y_test_removed, y_pred_removed_test)}')
    print('-----------------------------------------------------')


poly: 1
TRAIN -> mse: 84.98192996913082; r2: 0.556316099178253
TEST -> mse: 76.26670635430587; r2: 0.5439279212021337
-----------------------------------------------------
poly: 2
TRAIN -> mse: 68.21638170950628; r2: 0.643847693882533
TEST -> mse: 62.96371822820282; r2: 0.6234792974043474
-----------------------------------------------------
poly: 3
TRAIN -> mse: 61.39974063081097; r2: 0.6794368350728532
TEST -> mse: 69.32723534875986; r2: 0.5854257007516266
-----------------------------------------------------
poly: 4
TRAIN -> mse: 42.57017526465385; r2: 0.7777444990135207
TEST -> mse: 252.8344330225725; r2: -0.5119405435524531
-----------------------------------------------------
poly: 5
TRAIN -> mse: 8.40406097544144; r2: 0.956123065719005
TEST -> mse: 942623.6070497974; r2: -5635.8542518930035
-----------------------------------------------------
poly: 6
TRAIN -> mse: 1.1599372906883288e-23; r2: 1.0
TEST -> mse: 683467.6121893534; r2: -4086.11100272411
-----------------------------

generalmente il primo modello fitta molto meglio e overfitta meno.
Scelgo il modello con meno dati e provo ad applicare regolarizzazioni sul polinomio di 3 e 4 grado


In [17]:
from sklearn.linear_model import Ridge, Lasso, ElasticNet


In [18]:
alphas = [0.001, 0.01, 0.1, 1, 10]
polys = [2,3,4]

In [19]:
for poly in polys:
  print(f'poly: {poly}')
  pf = PolynomialFeatures(degree=poly)
  X_train_poly = pf.fit_transform(X_train_std)
  X_test_poly = pf.transform(X_test_std)
  
  for alpha in alphas:
    print(f'alpha: {alpha}')
    ridge = Ridge(alpha=alpha)
    ridge.fit(X_train_poly, y_train)
    
    y_pred_train_ridged = ridge.predict(X_train_poly)
    y_pred_test_ridged = ridge.predict(X_test_poly)
    
    print(f'TRAIN -> mse: {mean_squared_error(y_pred_train_ridged, y_train)}; r2: {r2_score(y_pred_train_ridged, y_train)}')
    print(f'TEST -> mse: {mean_squared_error(y_pred_test_ridged, y_test)}; r2: {r2_score(y_pred_test_ridged, y_test)}')
    print('-----------------------------------------------------')

poly: 2
alpha: 0.001
TRAIN -> mse: 53.994270264755706; r2: 0.6074321234576778
TEST -> mse: 57.878913361755465; r2: 0.5823328695528995
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 53.994283664926336; r2: 0.6073910058822966
TEST -> mse: 57.885659561303335; r2: 0.582248773319182
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 53.9955245047434; r2: 0.6069811062894732
TEST -> mse: 57.95016576172462; r2: 0.5814367101462119
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 54.06317925350774; r2: 0.6030692654200054
TEST -> mse: 58.376366221218895; r2: 0.5753657337316237
-----------------------------------------------------
alpha: 10
TRAIN -> mse: 54.98225915148856; r2: 0.5725413343979807
TEST -> mse: 58.638741143159336; r2: 0.5527094155765155
-----------------------------------------------------
poly: 3
alpha: 0.001
TRAIN -> mse: 39.26564408850715; r2: 0.7420247890804469
TEST -> mse: 80.74711179325672;

ridge
alpha 0.1 su poly 2
TRAIN -> mse: 53.994283664926336; r2: 0.6073910058822966
TEST -> mse: 57.885659561303335; r2: 0.582248773319182
alpha 1 su poly 3
TRAIN -> mse: 40.979846466914424; r2: 0.7211257315488906
TEST -> mse: 63.16678660651442; r2: 0.6199066966090391

qui performa meglio poly 3 alpha 1

In [20]:
for poly in polys:
  print(f'poly: {poly}')
  pf = PolynomialFeatures(degree=poly)
  X_train_poly = pf.fit_transform(X_train_std)
  X_test_poly = pf.transform(X_test_std)
  
  for alpha in alphas:
    print(f'alpha: {alpha}')
    lassoe = Lasso(alpha=alpha)
    lassoe.fit(X_train_poly, y_train)
    
    y_pred_train_lassoed = lassoe.predict(X_train_poly)
    y_pred_test_lassoed = lassoe.predict(X_test_poly)
    
    print(f'TRAIN -> mse: {mean_squared_error(y_pred_train_lassoed, y_train)}; r2: {r2_score(y_pred_train_lassoed, y_train)}')
    print(f'TEST -> mse: {mean_squared_error(y_pred_test_lassoed, y_test)}; r2: {r2_score(y_pred_test_lassoed, y_test)}')
    print('-----------------------------------------------------')

poly: 2
alpha: 0.001
TRAIN -> mse: 53.9947440142166; r2: 0.607117262957992
TEST -> mse: 57.871828035768445; r2: 0.5821144973897305
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 54.03181754228318; r2: 0.6040917999675128
TEST -> mse: 57.74758028902423; r2: 0.5804207117185705
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 55.18185336929032; r2: 0.5709407690533498
TEST -> mse: 56.48192264253834; r2: 0.5673622653346571
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 73.64080770235637; r2: 0.13469071528245802
TEST -> mse: 65.9598307239329; r2: 0.22720470337687404
-----------------------------------------------------
alpha: 10
TRAIN -> mse: 171.44094495466155; r2: -70.94045468121662
TEST -> mse: 148.97105139913978; r2: -42.63218064167871
-----------------------------------------------------
poly: 3
alpha: 0.001
TRAIN -> mse: 40.29703682868388; r2: 0.7326139746337348
TEST -> mse: 69.50432504177324; r

C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.201e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.323e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\loca

TRAIN -> mse: 22.503127596728657; r2: 0.8608456167807227
TEST -> mse: 609.9275818025717; r2: 0.12420739908476586
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 38.02086826735291; r2: 0.7293629911367163
TEST -> mse: 107.88323867420324; r2: 0.5398940910373609
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 66.46334239600576; r2: 0.2748496060659368
TEST -> mse: 64.15566451705995; r2: 0.31982006272853225
-----------------------------------------------------
alpha: 10
TRAIN -> mse: 150.9329367622686; r2: -4.402428370671203
TEST -> mse: 155.105678049015; r2: -1.245121341511299
-----------------------------------------------------


poly: 2
alpha: 0.001
TRAIN -> mse: 53.9947440142166; r2: 0.607117262957992
TEST -> mse: 57.871828035768445; r2: 0.5821144973897305
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 54.03181754228318; r2: 0.6040917999675128
TEST -> mse: 57.74758028902423; r2: 0.5804207117185705
-----------------------------------------------------
poly: 3
alpha: 0.001
TRAIN -> mse: 40.29703682868388; r2: 0.7326139746337348
TEST -> mse: 69.50432504177324; r2: 0.5802467269685756
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 41.20162520573907; r2: 0.7205712809071543
TEST -> mse: 64.58552137365746; r2: 0.5978612663645149
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 47.81316182236184; r2: 0.6415716537799765
TEST -> mse: 53.22182937823257; r2: 0.6113613869676937
-----------------------------------------------------


per ora il ridge sembra performare meglio

In [21]:
l1_ratios = [0.25, 0.5, 0.75]
for l1_ratio in l1_ratios:
  print(f'l1_ratio {l1_ratio}')
  for poly in polys:
    print(f'poly: {poly}')
    pf = PolynomialFeatures(degree=poly)
    X_train_poly = pf.fit_transform(X_train_std)
    X_test_poly = pf.transform(X_test_std)
    
    for alpha in alphas:
      print(f'alpha: {alpha}')
      lassoe = ElasticNet(alpha=alpha, l1_ratio=l1_ratio)
      lassoe.fit(X_train_poly, y_train)
      
      y_pred_train_lassoed = lassoe.predict(X_train_poly)
      y_pred_test_lassoed = lassoe.predict(X_test_poly)
      
      print(f'TRAIN -> mse: {mean_squared_error(y_pred_train_lassoed, y_train)}; r2: {r2_score(y_pred_train_lassoed, y_train)}')
      print(f'TEST -> mse: {mean_squared_error(y_pred_test_lassoed, y_test)}; r2: {r2_score(y_pred_test_lassoed, y_test)}')
      print('-----------------------------------------------------')

l1_ratio 0.25
poly: 2
alpha: 0.001
TRAIN -> mse: 54.00021381069504; r2: 0.6063730804560598
TEST -> mse: 58.02458132063833; r2: 0.5804024916710895
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 54.20600633472755; r2: 0.5976364704509043
TEST -> mse: 58.55629597880444; r2: 0.5701044166677312
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 56.50804792669307; r2: 0.5286514138455195
TEST -> mse: 58.51112389559937; r2: 0.5252994524599583
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 78.51504135854675; r2: -0.11772278257812063
TEST -> mse: 74.13176643742135; r2: 0.05332364362907138
-----------------------------------------------------
alpha: 10
TRAIN -> mse: 142.9901419118333; r2: -9.346507992043438
TEST -> mse: 125.44432311650729; r2: -6.031313183355875
-----------------------------------------------------
poly: 3
alpha: 0.001
TRAIN -> mse: 40.40626376524139; r2: 0.730092090764694
TEST -> mse: 67.72

C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.779e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.312e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\loca

TRAIN -> mse: 24.756585327710244; r2: 0.8433682722544311
TEST -> mse: 238.50528585270212; r2: 0.31899171889037437
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 35.97527245538182; r2: 0.7429001872203054
TEST -> mse: 207.10926613299685; r2: 0.3997740718604108
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 59.22420853750161; r2: 0.4265406010430771
TEST -> mse: 89.44857582021461; r2: 0.34537148867245626
-----------------------------------------------------
alpha: 10
TRAIN -> mse: 118.61765179122854; r2: -1.7054106294580609
TEST -> mse: 118.77621296067284; r2: -0.7923489301604829
-----------------------------------------------------
l1_ratio 0.5
poly: 2
alpha: 0.001
TRAIN -> mse: 53.99767295564826; r2: 0.6066188562582687
TEST -> mse: 57.97681016662907; r2: 0.5809448023704231
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 54.1423535474952; r2: 0.5996015130892226
TEST -> mse: 58.38176758754876; r2:

C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.180e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.104e+01, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\loca

TRAIN -> mse: 135.64104558815532; r2: -3.5516953943459777
TEST -> mse: 127.38864940236638; r2: -1.6846885271807288
-----------------------------------------------------
poly: 4
alpha: 0.001
TRAIN -> mse: 19.698345958303204; r2: 0.8818549976593981
TEST -> mse: 2098.689988016642; r2: -0.0014711135649092544
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 24.357311723557235; r2: 0.8465434533609982
TEST -> mse: 310.2287083039699; r2: 0.2508880886361605
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 36.77008866235079; r2: 0.7355352205289774
TEST -> mse: 155.29657704329335; r2: 0.4540861817963945
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 63.49994916854773; r2: 0.3408318363829206
TEST -> mse: 78.14006834377003; r2: 0.2826946767477565
-----------------------------------------------------
alpha: 10


C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.091e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.778e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\loca

TRAIN -> mse: 136.9856993731987; r2: -2.9150609303594157
TEST -> mse: 140.24580203418938; r2: -1.0291843893357897
-----------------------------------------------------
l1_ratio 0.75
poly: 2
alpha: 0.001
TRAIN -> mse: 53.995799779267394; r2: 0.6068669307506841
TEST -> mse: 57.92596391148562; r2: 0.5815146623679056
-----------------------------------------------------
alpha: 0.01
TRAIN -> mse: 54.07591527469423; r2: 0.6017488087511382
TEST -> mse: 58.10584274671212; r2: 0.5759614454550415
-----------------------------------------------------
alpha: 0.1
TRAIN -> mse: 55.58785545302476; r2: 0.5542959387551474
TEST -> mse: 57.74845970546234; r2: 0.5458874466293047
-----------------------------------------------------
alpha: 1
TRAIN -> mse: 77.89884830373654; r2: -0.026904118232181196
TEST -> mse: 70.42262801127184; r2: 0.12948651318701332
-----------------------------------------------------
alpha: 10
TRAIN -> mse: 167.90265864617652; r2: -40.380226668580576
TEST -> mse: 145.677610821685; r

C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.096e+02, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.002e+03, tolerance: 5.535e+00
  model = cd_fast.enet_coordinate_descent(
C:\Users\David\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\loca

attualmente le 3 possibilità su elasticsearch sono

l1_ratio 0.25
-----------------------------------------------------
poly: 3
alpha: 0.001
TRAIN -> mse: 40.40626376524139; r2: 0.730092090764694
TEST -> mse: 67.72366414554764; r2: 0.591080192922895
-----------------------------------------------------
l1_ratio 0.5
-----------------------------------------------------
poly: 3
alpha: 0.01
TRAIN -> mse: 41.886968089263725; r2: 0.710890793719221
TEST -> mse: 62.22677789105667; r2: 0.6147711779509234
-----------------------------------------------------
l1_ratio 0.75
-----------------------------------------------------
poly: 3
alpha: 0.01
TRAIN -> mse: 41.58034448712637; r2: 0.7149140799871441
TEST -> mse: 63.078031950347764; r2: 0.6088985288102959

direi che l1_ratio=0.5, alpha=0.01 e poly=3 mi pare un buon compromesso tra prestazioni e overfitting


avendo analizzato ridge, lasso ed eleasticsearch (con l1_ratio a 0.25, 0.5, 0.75) proponendo polinomi di grado da 1 a 10 e alpha tra 10^-4 e 10.
posso dichiarare che il miglior modello di machine learning è definito da
--------
algoritmo: ridge
alpha: 1
polinomio: 3

con le seguenti metriche di performance
--------

TRAIN -> mse: 40.979846466914424; r2: 0.7211257315488906
TEST -> mse: 63.16678660651442; r2: 0.6199066966090391